### DISCUSSION TITLE
How to change other input value properly

### DISCUSSION THREAD

**Discussion by pimmee** [react-hook-form/react-hook-form#1977] [Q&A] [answered]
Hello! First of all, thanks for a great package. I really like the API.

I have a few questions.

1. What's the best way to change an other input's value? I've tried doing it in onChange / validate, but don't know if this is correct.
2. How can you invalidate an input field, without showing error message? My use case is that I conditionally display an input, and I want the `isValid` state to turn false if this input should be displayed, but I don't want to trigger a validation because this displays an error message.
3. Why does 2 renders occur when I enter a letter into the textbox?
4. Why does 4 renders occur when I enter a value into radio buttons?

I have a pretty complex form that I'm working on, and the number of renders with this package is only really decreasing compared to controlled validation when entering long texts into text inputs, so I must be doing something horribly wrong! Some pointers would be really appreciated!

Here's a [codesandbox](https://codesandbox.io/s/elated-davinci-8yjn8?file=/src/App.js)

**Comment by keiya01** [comment, marked as answer]
Thank you for using react-hook-form.

About 1, 2:

1. I think you don't need to invoke `setValue` in `roles` props.(https://codesandbox.io/s/reverent-proskuriakova-5u4sx)
2. I think this issue is resolved by above solution

**Reply by pimmee** [reply]
But if you click "Yes" on `fieldOne`, and then any value on `fieldTwo`, then select "No" and then "Yes" again  on `fieldTwo` - the previous value is still selected? I want it to be cleared.
Thank you

**Reply by keiya01** [reply]
I see... Fixed it. https://codesandbox.io/s/reverent-proskuriakova-5u4sx?file=/src/App.js

**Reply by keiya01** [reply]
About 3, 4:

These re-render occur when `watch`, `validation` process, `setValue` are invoked.

**Reply by keiya01** [reply]
I think I could resolve your issue but this solution is not best.
@bluebill1049 Please check this issue when you have times.

**Reply by pimmee** [reply]
Hmm this seems like a pretty common use case, with values in a form affecting other values. I wish there was more examples showcasing a bit more complex forms. For example, the current [example](https://github.com/react-hook-form/react-hook-form/blob/master/examples/setValue.tsx) for setting value is by clicking a button which doesn't feel like a real world example.

 Even when not using`formState`, 2 renders occur when I enter a letter. By reading the docs my impression is that rerenders should only occur when:

- a) A variable that you "watch" change
- b) An error has occurred

With this in mind, I actually don't even understand why even one render should occur.

Grateful for help!

**Reply by bluebill1049** [reply]
> Even when not using`formState`, 2 renders occur when I enter a letter. By reading the docs my impression is that rerenders should only occur when:

1. you are watching a value: first render.
2. you are watching `formState`, why `formState` trigger a separate re-render? because validation can be async

**Reply by pimmee** [reply]
I have removed `watch` and `formState` from [this example](https://codesandbox.io/s/intelligent-paper-nfkzv?file=/src/App.js) and 2 renders still occur whenever I enter a letter into the text input. Does 2 renders occur simply if an input is required in `onChange` mode?
This causes 2 renders when I enter a letter:
```
<input name="name" ref={register({ required: "Error name" })} />
```
And this causes 0 renders:
```
<input name="name" ref={register} />
```
I'm not watching any value or `formState` in this example and this still occur.
```
  const { handleSubmit, register, getValues } = useForm({
    mode: "onChange",
    reValidateMode: "onChange",
    defaultValues: {
      name: "",
    }
  });
```

**Reply by pimmee** [reply]
Should I create an issue for this? Or is this the expected behavior? 

Thank you.

**Reply by bluebill1049** [reply]
Ok, take a closer look I see the reason, this not a bug. This is due to `isValid`.

because react hook form is uncontrolled which means we wouldn't know the input value until the render phase, so the first phase we have to assume the form is true, then set isValid to false (if any) and skip the rest of the re-rendering. if we still initial to `false`, then it will trigger multiple re-render, because then it will need to set to `true` and then next input `false` again and etc. hope this make a bit more sense.

**Reply by pimmee** [reply]
Intuitively I don't understand why it should happen, but I'm sure you know what you're talking about. Unfortunately `react-hook-form` is not fitting for my use case then. Thanks for the quick responses!
